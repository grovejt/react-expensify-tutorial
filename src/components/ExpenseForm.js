import React, { useState } from "react";
import moment from "moment";
import "react-dates/initialize";
import { SingleDatePicker } from "react-dates";

const ExpenseForm = props => {
  // State:
  const [expense, setExpense] = useState({
    description: props.expense ? props.expense.description : "",
    amount: props.expense ? (props.expense.amount / 100).toString() : "",
    note: props.expense ? props.expense.note : ""
    //createdAt: props.expense ? moment(props.expense.createdAt) : moment()
  });
  // must break date into separate state due to buggy airbnb date picker:
  const [calDate, setCalDate] = useState(props.expense ? moment(props.expense.createdAt) : moment());
  const [calendarFocused, setCalendarFocused] = useState(false);
  const [error, setError] = useState(false);

  // Handlers:
  const onDescriptionChange = e => {
    setExpense({
      ...expense,
      description: e.target.value
    });
  };

  const onAmountChange = e => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      setExpense({
        ...expense,
        amount: amount
      });
    }
  };

  const onDateChange = createdAt => {
    console.log("onDateChange", createdAt);
    if (createdAt) {
      setCalDate(createdAt);
    }
    // setExpense({
    //   ...expense,
    //   createdAt
    // });
  };

  const onCalendarFocusChange = ({ focused }) => {
    console.log("onCalendarFocusChange", focused, calendarFocused);

    setCalendarFocused(focused);
  };

  const onNoteChange = e => {
    setExpense({
      ...expense,
      note: e.target.value
    });
  };

  const onSubmit = e => {
    e.preventDefault();
    if (!expense.description || !expense.amount) {
      setError(true);
    } else {
      setError(false);
      const formattedAmount = parseFloat(expense.amount, 10) * 100;
      const timestamp = calDate.valueOf();
      props.onSubmit({
        ...expense,
        amount: formattedAmount,
        createdAt: timestamp
      });
    }
  };

  return (
    <div>
      {error ? <p>Please enter both a description and an amount!</p> : ""}

      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="Description"
          autoFocus
          value={expense.description}
          onChange={onDescriptionChange}
        />
        <input type="text" placeholder="Amount" value={expense.amount} onChange={onAmountChange} />

        <SingleDatePicker
          id="your_unique_id" // PropTypes.string.isRequired,
          date={calDate} // momentPropTypes.momentObj or null
          onDateChange={onDateChange} // PropTypes.func.isRequired
          focused={calendarFocused} // PropTypes.bool
          onFocusChange={onCalendarFocusChange} // PropTypes.func.isRequired
          numberOfMonths={1}
          isOutsideRange={() => false}
        />

        <br />
        <textarea placeholder="Add a note for your expense (optional)" value={expense.note} onChange={onNoteChange} />
        <button className="btn btn-primary">Add Expense</button>
      </form>
    </div>
  );
};

export default ExpenseForm;
