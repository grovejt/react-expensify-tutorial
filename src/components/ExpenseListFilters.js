import React, { useState } from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from "react-dates";

const ExpenseListFilters = props => {
  //State
  const [calendarFocused, setCalendarFocused] = useState(null);

  const onDatesChange = ({ startDate, endDate }) => {
    props.setStartDate(startDate);
    props.setEndDate(endDate);
  };

  const onCalendarFocusChange = focused => {
    //console.log("onCalendarFocusChange", focused, calendarFocused);
    setCalendarFocused(focused);
  };

  return (
    <div>
      <p>expense list filters</p>
      <input
        type="text"
        value={props.filters.text}
        onChange={e => {
          props.setTextFilter(e.target.value);
        }}
      />
      <span>Sort By:</span>
      <select
        value={props.filters.sortBy}
        onChange={e => {
          // console.log(e.target.value);
          if (e.target.value === "date") {
            props.sortByDate();
          } else if (e.target.value === "amount") {
            props.sortByAmount();
          }
        }}
      >
        <option value="date">Date</option>
        <option value="amount">Amount</option>
      </select>

      <DateRangePicker
        startDate={props.filters.startDate} // momentPropTypes.momentObj or null,
        startDateId="your_unique_start_date_id" // PropTypes.string.isRequired,
        endDate={props.filters.endDate} // momentPropTypes.momentObj or null,
        endDateId="your_unique_end_date_id" // PropTypes.string.isRequired,
        onDatesChange={onDatesChange} // PropTypes.func.isRequired,
        focusedInput={calendarFocused} // PropTypes.oneOf([START_DATE, END_DATE]) or null,
        onFocusChange={onCalendarFocusChange} // PropTypes.func.isRequired,
        showClearDates={true}
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
    </div>
  );
};

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

const mapDispatchToProps = { setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate };

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
