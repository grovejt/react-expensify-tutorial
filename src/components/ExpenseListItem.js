import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";
import { startRemoveExpense } from "../actions/expenses";

export const ExpenseListItem = ({ id, description, amount, createdAt, startRemoveExpense }) => {
  return (
    <Link className="list-item" to={`/edit/${id}`}>
      <div>
        <h3 className="list-item__title">{description}</h3>
        <span className="list-item__sub-title">{moment(createdAt).format("MMMM Do, YYYY")}</span>
      </div>
      <h3 className="list-item__data">{numeral(amount / 100).format("$0,0.00")}</h3>
    </Link>
  );
};

const mapDispatchToProps = { startRemoveExpense };
export default connect(null, mapDispatchToProps)(ExpenseListItem);

{
  /*  
export const ExpenseListItemOld = ({ id, description, amount, createdAt, startRemoveExpense }) => {
  return (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
        <p>
          {numeral(amount / 100).format("$0,0.00")}-{moment(createdAt).format("MMMM Do, YYYY")}
        </p>
      </Link>

      <button
        className="btn btn-primary"
        onClick={e => {
          startRemoveExpense(id);
        }}
      >
        Remove
      </button>
    </div>
  );
};

*/
}
