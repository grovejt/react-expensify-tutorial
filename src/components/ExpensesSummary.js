import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import numeral from "numeral";
import getExpensesTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";

export const ExpenseSummary = ({ expenseCount, expenseTotal }) => {
  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          Viewing <span>{expenseCount} </span> expense{expenseCount === 1 ? "" : "s"} <span> totalling </span>
          {numeral(expenseTotal / 100).format("$0,00.00")}
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">
            Add Expense
          </Link>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: getExpensesTotal(visibleExpenses)
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
