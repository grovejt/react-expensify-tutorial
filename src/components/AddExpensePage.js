import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
// import PropTypes from "prop-types";
import { startAddExpense } from "../actions/expenses";

export const AddExpensePage = props => {
  const addExpense = expense => {
    props.startAddExpense(expense);
    props.history.push("/");
  };
  return (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm onSubmit={addExpense} />
    </div>
  );
};

const mapStateToProps = state => {
  return {
    expenses: state.expenses
  };
};

const mapDispatchToProps = { startAddExpense };

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);
