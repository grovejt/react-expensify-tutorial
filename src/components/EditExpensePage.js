import React from "react";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export const EditExpensePage = props => {
  const updateExpense = updatedExpense => {
    // console.log("submitted", updatedExpense);
    // console.log("submitted", { ...updatedExpense });
    props.editExpense(props.expense.id, updatedExpense);
    props.history.push("/");
  };
  return (
    <div>
      <h1>Edit Expense</h1>
      <ExpenseForm expense={props.expense} onSubmit={updateExpense} />
      <button
        className="btn btn-primary"
        onClick={e => {
          props.removeExpense(props.expense.id);
          props.history.push("/");
        }}
      >
        Remove{" "}
      </button>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find(expense => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = { editExpense, removeExpense };

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
