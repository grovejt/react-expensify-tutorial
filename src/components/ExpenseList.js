import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = props => {
  const expenses = props.expenses;
  return (
    <div>
      <h1>Expense List</h1>
      <ul>
        {expenses.map(expense => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        })}
      </ul>
    </div>
  );
};

const mapStateToProps = state => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
