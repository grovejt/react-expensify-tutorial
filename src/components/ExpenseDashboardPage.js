import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseTotals from "./ExpensesSumary";
import ExpenseListFilters from "./ExpenseListFilters";

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseTotals />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;
