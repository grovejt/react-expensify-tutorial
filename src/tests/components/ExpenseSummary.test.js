import React from "react";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { ExpenseSummary } from "../../components/ExpensesSummary";

// test("should render Expense Summary correctly", () => {
// const wrapper = shallow(<ExpensesSumary />);
// expect(wrapper).toMatchSnapshot();
// });

configure({ adapter: new Adapter() });

describe("<ExpensesSumary/>", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<ExpenseSummary expenseCount={2} expenseTotal={111} />);
  });
  it("should render <ExpensesSumary /> element", () => {
    expect(wrapper).toMatchSnapshot();
    // expect(wrapper.find(Link)).toHaveLength(1);
  });
});
