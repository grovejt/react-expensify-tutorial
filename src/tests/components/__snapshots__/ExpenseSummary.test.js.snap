// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ExpensesSumary/> should render <ExpensesSumary /> element 1`] = `
<div>
  <h1>
    Viewing 
    2
     expense
    s
     totalling
     
    $1.11
  </h1>
</div>
`;
